### PR TITLE
Created methods for DB and validation

### PR DESCRIPTION
**Pull request by larinpaul**

### CODE REVIEW

File: src/main/java/com/example/employeecrudapplication/service/EmployeeService.java [hunk 1]
```diff
@@ -1,35 +1,74 @@
...
 import org.springframework.stereotype.Service;
-import org.springframework.web.bind.annotation.RequestBody;
+
+import java.util.ArrayList;
+import java.util.List;
+import java.util.Optional;
 
 @Service
 @RequiredArgsConstructor
 public class EmployeeService {
```
Comment (sorokaa) [inline src/main/java/com/example/employeecrudapplication/service/EmployeeService.java]: Use interface for inversion of control

File: src/main/java/com/example/employeecrudapplication/service/EmployeeService.java [hunk 2]
```diff
@@ -1,35 +1,74 @@
...
-    public Employee createEmployee(EmployeeDto employee) {
-        validator.validToCreate(employee);
-        // Сделать Маппер.
-        // Берешь EmployeeDto, создаешь объект класса Employee
-        // Прокидываешь поля из DTO в Entity, то есть из EmployeeDto берём поля, и каждое поле маппим в Employee
-        // На выходе получится объект класса Employee, который сохраняешь в базу данных
+        List<EmployeeDto> employeesDto = new ArrayList<>();
 
-        return this.employeeRepository.save(employee);
+        for (Employee employee : all) {
```
Comment (sorokaa) [inline src/main/java/com/example/employeecrudapplication/service/EmployeeService.java]: Use stream API

File: src/main/java/com/example/employeecrudapplication/service/EmployeeService.java [hunk 3]
```diff
@@ -1,35 +1,74 @@
...
+            EmployeeDto employeeDto = employeeMapper.toDto(employee);
+            employeesDto.add(employeeDto);
+        } // Вообще это можно сделать в одну конструкцию, использовав streamAPI (можно поискать потом)
+        return employeesDto;
     }
 
+    public EmployeeDto findById(Long employeeId) {
 
+        Optional<Employee> byId = employeeRepository.findById(employeeId);
```
Comment (sorokaa) [inline src/main/java/com/example/employeecrudapplication/service/EmployeeService.java]: Use methods from Optional (map, orElseThrow)

File: src/main/java/com/example/employeecrudapplication/validator/EmployeeValidator.java [hunk 4]
```diff
@@ -0,0 +1,45 @@
...
+
+    public void validToCreate(EmployeeDto employee) {
+        validateFields(employee);
+    }
+
+    public void validToUpdate(EmployeeDto employee) {
+        validateFields(employee);
+    }
+
+//    public void validForUpdate() {
```
Comment (sorokaa) [inline src/main/java/com/example/employeecrudapplication/validator/EmployeeValidator.java]: Remove

File: src/main/java/com/example/employeecrudapplication/validator/EmployeeValidator.java [hunk 5]
```diff
@@ -0,0 +1,45 @@
...
+        }
+        validateEmail(employee.getEmail());
+    }
+
+    public void validToCreate(EmployeeDto employee) {
+        validateFields(employee);
+    }
+
+    public void validToUpdate(EmployeeDto employee) {
+        validateFields(employee);
```
Comment (sorokaa) [inline src/main/java/com/example/employeecrudapplication/validator/EmployeeValidator.java:48]: Add email validation

File: src/main/java/com/example/employeecrudapplication/validator/EmployeeValidator.java [hunk 6]
```diff
@@ -0,0 +1,45 @@
...
+import com.example.employeecrudapplication.exception.EmployeeValidationException;
+import com.example.employeecrudapplication.model.EmployeeDto;
+import org.apache.commons.lang3.StringUtils;
+import org.springframework.stereotype.Component;
+
+@Component
+public class EmployeeValidator {
+
+    public void validateEmail(String email) {
+        if (
```
Comment (sorokaa) [inline src/main/java/com/example/employeecrudapplication/validator/EmployeeValidator.java]: Make normal email validation (e.g., regex)

File: src/main/java/com/example/employeecrudapplication/validator/EmployeeValidator.java [hunk 7]
```diff
@@ -0,0 +1,45 @@
...
+                || (!email.contains("@"))
+                || (!email.contains("."))
+        ) {
+            throw new EmployeeValidationException("Employee Email is not present in the request");
+        }
+    }
+
+    public void validateFields(EmployeeDto employee) {
+        if (employee == null) {
+            throw new EmployeeValidationException("Null Pointer Exception! Employee request is null");
```
Comment (sorokaa) [inline src/main/java/com/example/employeecrudapplication/validator/EmployeeValidator.java]: Change message to "Employee not provided"

File: src/main/java/com/example/employeecrudapplication/validator/EmployeeValidator.java [hunk 8]
```diff
@@ -0,0 +1,45 @@
...
+@Component
+public class EmployeeValidator {
+
+    public void validateEmail(String email) {
+        if (
+                StringUtils.isBlank(email)
+                || (!email.contains("@"))
+                || (!email.contains("."))
+        ) {
+            throw new EmployeeValidationException("Employee Email is not present in the request");
```
Comment (sorokaa) [inline src/main/java/com/example/employeecrudapplication/validator/EmployeeValidator.java]: Make here two validations:
1. If email is empty, then message should be "Email isn't provided"
2. If email is invalid, then message should be "Invalid email"

File: src/main/java/com/example/employeecrudapplication/validator/EmployeeValidator.java (hunk 7 above)
Comment (sorokaa) [inline src/main/java/com/example/employeecrudapplication/validator/EmployeeValidator.java]: For EmployeeValidationException make response status 400 BAD REQUEST

File: src/main/java/com/example/employeecrudapplication/validator/EmployeeValidator.java [hunk 9]
```diff
@@ -0,0 +1,45 @@
...
+
+import com.example.employeecrudapplication.exception.EmployeeValidationException;
+import com.example.employeecrudapplication.model.EmployeeDto;
+import org.apache.commons.lang3.StringUtils;
+import org.springframework.stereotype.Component;
+
+@Component
+public class EmployeeValidator {
+
+    public void validateEmail(String email) {
```
Comment (sorokaa) [inline src/main/java/com/example/employeecrudapplication/validator/EmployeeValidator.java:19]: Make validation that empoyee with provided email isn't exists (see custom methods for JpaRepository)

**Review by sorokaa — CHANGES_REQUESTED** [review]

File: src/main/java/com/example/employeecrudapplication/validator/EmployeeValidator.java (hunk 7 above)
Comment (larinpaul) [reply src/main/java/com/example/employeecrudapplication/validator/EmployeeValidator.java]: Tried doing it like this, is this correct?
image

Comment (sorokaa) [reply src/main/java/com/example/employeecrudapplication/validator/EmployeeValidator.java]: @larinpaul https://www.baeldung.com/spring-response-status-exception

File: src/main/java/com/example/employeecrudapplication/service/EmployeeService.java (hunk 3 above)
Comment (larinpaul) [reply src/main/java/com/example/employeecrudapplication/service/EmployeeService.java]: Trying to change the method, don't know which way is the best
image

File: src/main/java/com/example/employeecrudapplication/validator/EmployeeValidator.java (hunk 4 above)
Comment (larinpaul) [reply src/main/java/com/example/employeecrudapplication/validator/EmployeeValidator.java]: Removed, now it's like this:
image

File: src/main/java/com/example/employeecrudapplication/validator/EmployeeValidator.java (hunk 8 above)
Comment (larinpaul) [reply src/main/java/com/example/employeecrudapplication/validator/EmployeeValidator.java]: image

File: src/main/java/com/example/employeecrudapplication/validator/EmployeeValidator.java (hunk 7 above)
Comment (larinpaul) [reply src/main/java/com/example/employeecrudapplication/validator/EmployeeValidator.java]: image

File: src/main/java/com/example/employeecrudapplication/service/EmployeeService.java (hunk 3 above)
Comment (sorokaa) [reply src/main/java/com/example/employeecrudapplication/service/EmployeeService.java]: Use map method for mapping employee to EmployeeDto

File: src/main/java/com/example/employeecrudapplication/validator/EmployeeValidator.java (hunk 7 above)
Comment (sorokaa) [reply src/main/java/com/example/employeecrudapplication/validator/EmployeeValidator.java]: Method should be void. Thus, remove return statement

File: src/main/java/com/example/employeecrudapplication/validator/EmployeeValidator.java (hunk 6 above)
Comment (larinpaul) [reply src/main/java/com/example/employeecrudapplication/validator/EmployeeValidator.java]: Changed validateEmail and created a method patternMatches (is it a helper method?)
image

File: src/main/java/com/example/employeecrudapplication/validator/EmployeeValidator.java (hunk 7 above)
Comment (larinpaul) [reply src/main/java/com/example/employeecrudapplication/validator/EmployeeValidator.java]: I don't understand what I should do yet, but I have just made it like this:
image
I will try reading https://www.baeldung.com/spring-response-status-exception as you suggested

File: src/main/java/com/example/employeecrudapplication/validator/EmployeeValidator.java (hunk 6 above)
Comment (sorokaa) [reply src/main/java/com/example/employeecrudapplication/validator/EmployeeValidator.java]: 1. Variable at line 25 is redundant
2. You call method _patternMatches_ twice (line 28 and 29)
3. You could make patternMatches method private and non-static

File: src/main/java/com/example/employeecrudapplication/validator/EmployeeValidator.java (hunk 5 above)
Comment (larinpaul) [reply src/main/java/com/example/employeecrudapplication/validator/EmployeeValidator.java:48]: I don't understand how I can validate the email String. Should I inject employeeRepository and use the get method? Or should I do something else?
image

Comment (sorokaa) [reply src/main/java/com/example/employeecrudapplication/validator/EmployeeValidator.java:48]: Ok, you already have email validation inside validateFields method. Let's ignore this comment

File: src/main/java/com/example/employeecrudapplication/validator/EmployeeValidator.java (hunk 6 above)
Comment (larinpaul) [reply src/main/java/com/example/employeecrudapplication/validator/EmployeeValidator.java]: 1. Got rid of the variable
2. Removed the unnecessary method call
3. I've made patternMatches private and non-static, but now I don't understand how to use .patternMatches. The IDE suggests that I could make it static to solve this problem. What should I do?

Comment (sorokaa) [reply src/main/java/com/example/employeecrudapplication/validator/EmployeeValidator.java]: Call that method without **EmployeeValidator.**, just like plain method

File: src/main/java/com/example/employeecrudapplication/validator/EmployeeValidator.java (hunk 7 above)
Comment (larinpaul) [reply src/main/java/com/example/employeecrudapplication/validator/EmployeeValidator.java]: I don't understand how to make a solution for this. This issue will be reserved for a Skype call

File: src/main/java/com/example/employeecrudapplication/validator/EmployeeValidator.java (hunk 6 above)
Comment (larinpaul) [reply src/main/java/com/example/employeecrudapplication/validator/EmployeeValidator.java]: Done
image

Comment (sorokaa) [reply src/main/java/com/example/employeecrudapplication/validator/EmployeeValidator.java]: To follow the best practices, move the private method to the bottom of this class

**Review by sorokaa — APPROVED** [review]

**Pull request merged**